pkg/tlsutil/tls.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsutil
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CertType defines the type of the cert.
+type CertType int
+
+const (
+	// ClientCert defines a client cert.
+	ClientCert CertType = iota
+	// ServingCert defines a serving cert.
+	ServingCert
+	// ClientAndServingCert defines both client and serving cert.
+	ClientAndServingCert
+)
+
+// CertConfig configures how to generate the Cert.
+type CertConfig struct {
+	// CertName is the name of the cert.
+	CertName string
+	// Optional CertType. Serving, client or both; defaults to both.
+	CertType CertType
+	// Optional CommonName is the common name of the cert; defaults to "".
+	CommonName string
+	// Optional Organization is Organization of the cert; defaults to "".
+	Organization []string
+	// Optional CA Key, if user wants to provide custom CA key via a file path.
+	CAKey string
+	// Optional CA Certificate, if user wants to provide custom CA cert via file path.
+	CACert string
+	// TODO: consider to add passed in SAN fields.
+}
+
+// CertGenerator is an operator specific TLS tool that generates TLS assets for the deploying a user's application.
+type CertGenerator interface {
+	// GenerateCert generates a secret containing TLS encryption key and cert, a Secret
+	// containing the CA key, and a ConfigMap containing the CA Certificate given the Custom
+	// Resource(CR) "cr", the Kubernetes Service "Service", and the CertConfig "config".
+	//
+	// GenerateCert creates and manages TLS key and cert and CA with the following:
+	// CA creation and management:
+	// - If CA is not given:
+	//  - a unique CA is generated for the CR.
+	//  - CA's key is packaged into a Secret as shown below.
+	//  - CA's cert is packaged in a ConfigMap as shown below.
+	//  - The CA Secret and ConfigMap are then put into the Cluster before returning then to
+	//    the user.
+	// - If CA is given:
+	//  - CA's key is packaged into a Secret as shown below.
+	//  - CA's cert is packaged in a ConfigMap as shown below.
+	//  - The CA Secret and ConfigMap are returned to the user without putting them into the
+	//    cluster; the reason behind this is that we don't know how to manage the passed in
+	//    CA. Hence, we simply return them as in Kubernetes Objects, and let the user decides
+	//    what to do such as whether to put them into kubernte or not.
```

Comment:
How about something a little shorter:
```
The CA Secret and Configmap are returned but not created on the k8s cluster. The operator does not manage the user specified CA.
```

Reply:
we can, but i think we should still add a why component.

Reply:
Alright how about:
```
// The CA Secret and Configmap are returned but not created on the k8s cluster. 
// The operator does not manage the CA because the user controls the lifecycle of the CA.
```

Reply:
sounds better.

Reply:
do you mean that `GenerateCert does not manage`?

Reply:
Yeah the CertGenerator is more specific. Even though it is part of the operator.